prime-router/docs/proposals/0017-metadata-tables-cleanup.md

```diff
@@ -0,0 +1,441 @@
+
+## Requirements
+
+- The new table structure must be able to capture incoming ELR messages of any type, not just SARS-CoV-2 or hMPXv
+- It must be flexible enough to accept incomplete data because we receive data that is
+    - Deidentified
+    - Low or very low quality, for example OTC records
+    - Non-standardized, for example, not all jurisdictions or senders will use correct or standardized values for CE and CWE segments
+    - From non-US jurisdictions
+    - Non-human data (potential future use case)
+- Minimize the amount of duplication to the barest minimum
+- Maximize our ability to use indexing, foreign keys, etc
+- Represent the hierarchical structure of HL7 and FHIR ELR messages
+- Not impact the existing codebase or Metabase in any way
+- Have near-zero disruption when deployed
+
+## Proposed Solution
+
+Please refer to the below diagram which shows a proposed reorganization of the data that currently exists in the CovidResultMetadata and ElrResultMetadata tables.
+
+![](0017-metadata-tables-cleanup/proposed-metadata-restructuring.png)
+
+There are a few pieces to this work that I will briefly outline here and then go into deeper detail. There are two stages to this work, one set that is immediate, and then a second stage that will come later. They are identified as `Phase 1` and `Phase 2`.
+
+### Solution Outline
+
+#### Phase 1
+This work will happen almost immediately in an effort to prepare for the Universal Pipeline
+
+1. Create the agreed upon table structure that will meet the requirements above
+2. Create a stored procedure that takes the data it is given, and then handles the separation and normalization of the data as needed
+    - This stored procedure will in the short term do the insert into the existing tables as well, so there's no disruption to existing the existing process. 
+    - Create a flag in the database as a configuration setting that controls whether to insert into the legacy table or not 
+3. Create a script that will migrate existing legacy data from the `covid_result_metadata` and `elr_result_metadata` tables into the new table structure
+4. Create two materialized views, one named `covid_result_metadata` and one named `elr_result_metadata` that will exactly replicate the layout of the current tables in order to prevent breakage of any dashboards in Metabase or existing queries
+    - As part of this, the two existing tables will be renamed but preserved for some period of time until we can be sure we have not lost any data
+
+#### Phase 2
+This work will happen as a second step, hopefully in Q1 2023, and reflect a move towards more independent microservices. On ingestion, ReportStream will invoke a separate Azure function that will queue up messages to process, which will then be saved into our metadata tables. The goal is to separate the concerns of processing from the goal of tracking metadata from the messages we receive.
```

Comment:
Love this idea of moving towards a microservices architecture.  Perhaps, we can take this chance to refactor and introduce the SOLID Principle (Single Responsibility, Open and Closed, Liskov Substitution, Interface Segregation, Dependency Inversion) into our class implementation.   Also, we need to update most of our unit and integration tests as well.